xarray/backends/api.py

```diff
@@ -1827,18 +1842,242 @@ def to_zarr(
                     "mode='r+'. To allow writing new variables, set mode='a'."
                 )
 
+    return write_to_zarr_store(
+        dataset,
+        zstore,
+        encoding,
+        compute,
+        chunkmanager_store_kwargs,
+        encode_coordinates=True,
+    )
+
+
+def write_to_zarr_store(
+    dataset, store, encoding, compute, chunkmanager_store_kwargs, encode_coordinates
+):
     writer = ArrayWriter()
     # TODO: figure out how to properly handle unlimited_dims
-    dump_to_store(dataset, zstore, writer, encoding=encoding)
+    dump_to_store(
+        dataset, store, writer, encoding=encoding, encode_coordinates=encode_coordinates
+    )
     writes = writer.sync(
         compute=compute, chunkmanager_store_kwargs=chunkmanager_store_kwargs
     )
 
     if compute:
-        _finalize_store(writes, zstore)
+        _finalize_store(writes, store)
     else:
         import dask
 
-        return dask.delayed(_finalize_store)(writes, zstore)
+        return dask.delayed(_finalize_store)(writes, store)
+    return store
+
+
+def initialize_zarr(
+    ds: Dataset,
+    store: MutableMapping,
+    *,
+    region_dims: Iterable[Hashable] = tuple(),
+    mode: Literal["w", "w-"] = "w-",
+    zarr_version: int | None = None,
+    consolidated: bool | None = None,
+    encoding: dict | None = None,
+    **kwargs,
+) -> Dataset:
+    """
+    Initialize a Zarr store with metadata.
+
+    This function initializes a Zarr store with all indexed coordinate variables, and
+    metadata for every variable in the dataset.
+    If ``region_dims`` is specified, it will also
+      1. Write variables that don't have any of ``region_dims``, and
+      2. Return a dataset with the remaining variables, which contain one or more of ``region_dims``.
```

Comment:
```suggestion
      2. Return a dataset with the remaining variables and indexes, which contain one or more of ``region_dims``.
```

...is that right?

Should we drop the non-region indexes?

My mental model of this (but it's from a few months ago):
- For indexes that are part of the region, we want to keep them now that we have `region="auto"`. When we use `region="auto"`, it won't rewrite those indexes
- For other indexes, we want to drop them, because otherwise each process will attempt to write them, and this can lead to conflicts

Reply:
> When we use region="auto" in each process, it correctly won't rewrite those indexes

Surprising to me, but you are correct! https://github.com/pydata/xarray/issues/8589

Currently we keep any indexes needed for any variable that has dimensions that overlap with `region_dims`. So for the test case, there is a variable with dims `(x, y)`. If `region_dims==("x",)`, then the index `y` is also returned. This works fine with `region="auto"` but requires adding `"y": slice(None)` when explicitly specifying `region`.

> For other indexes, we want to drop them, because otherwise each process will incorrectly attempt to write them, and this can lead to conflicts

We do end up dropping those indexes. The test case has a variable with dimensions `("z",)` but the regions only contain `"x"` and/or `"y"`. The `z` index ends up getting dropped.